components/internal/typescript-base-compiler/typescript-base.js

```diff
@@ -0,0 +1,100 @@
+    const nonCompiled = !filesByToCompile.false ? [] : filesByToCompile.false.map(file => _getDistFile(file, distPath, false));
+    return compiled.concat(nonCompiled);
+  }
+  
+  let oldProgram = undefined
+  function compileComponent(files, options, path, context) {
+    
+      let program = ts.createProgram(files.map(file => file.path), options, undefined, oldProgram);
+      oldProgram = program
+      const sources = program.getSourceFile(context.componentObject.mainFile)
+      
+      let declarationFile;
+      program.emit(sources, (file, data) => {
+          declarationFile = new Vinyl({
+              contents: new Buffer(data),
+              base: path,
+              path: _getDistFilePath({relative:file.split(longestStartingSubstring(file, path))[1]}, path, true)
+              .replace('d.js', 'd.ts'),
+          });
+      },undefined, true);
+      
+      const results = files.map(file => {
+        if (isDefJS(file)) {
+          return null
+        }
+          const result = ts.transpileModule(file.contents.toString(), 
+              { 
+                  compilerOptions: options.compilerOptions, 
+                  fileName: file.basename, 
+                  moduleName: file.basename.split('.')[0]
+              });
+          const mappings = new Vinyl({
+              contents: new Buffer(result.sourceMapText),
+              base: path,
+              path: _getDistFilePath(file, path, true),
+          });
+          mappings.basename = _getRevisedFileExtension(file.basename) + '.map';
+          const fileContent = result.outputText ?  new Buffer(`${result.outputText}\n\n//# sourceMappingURL=${result.sourceMapText}`) : new Buffer(result.outputText);
+          const distFile = _getDistFile(file, path, true, fileContent);
+          return [mappings, distFile, declarationFile];
+      })
+      .filter((a)=> !!a)
+      .reduce((a, b) => a.concat(b), [declarationFile])
+  
+      return results
+  }
+  const _toCompile = (compiledFileTypes) => (file) => {
+      return compiledFileTypes.indexOf(file.extname.replace('.','')) > -1 && !file.stem.endsWith('.d');
+  }
+  
+  const _getDistFile = (file, distPath, reviseExtension, content) => {
+      let distFile = file.clone();
+      distFile.base = distPath;
+      distFile.path = _getDistFilePath(file, distPath, reviseExtension);
+  
+      if (content) {
+          distFile.contents = content;
+      }
+  
+      return distFile;
+  }
+  
+  const _getDistFilePath = (file, distPath, reviseExtension) => {
+      var fileRelative = file.relative;
+  
+      if (reviseExtension) fileRelative = _getRevisedFileExtension(file.relative);
+  
+      return path.join(distPath, fileRelative);
+  }
+  
+  const _getRevisedFileExtension = (fileName) => {
+      return fileName.replace('.tsx', '.js').replace('.ts', '.js');
+  }
+  
+  function longestStartingSubstring(left, right) {
```

Comment:
There are some nice implementation of this described here:
https://stackoverflow.com/questions/1916218/find-the-longest-common-starting-substring-in-a-set-of-strings
We also implemented it in bit (it should be a component of course)
https://github.com/teambit/bit/blob/master/src/utils/array/shared-start.js
(I don't expect you to change it.. just to let you know about it)
but a comment about what it does will be nice